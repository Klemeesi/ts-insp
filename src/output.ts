import { ImportInfo } from "./types";

export const exportToConsole = (imports: ImportInfo[]) => {
  imports.forEach((i) => {
    const indentation = " ".repeat((i.level - 1) * 2);
    const branch = i.level > 1 ? " â”” " : " - ";
    console.log(`${indentation}${branch}${i.import}`);

    exportToConsole(i.imports);
  });
};
