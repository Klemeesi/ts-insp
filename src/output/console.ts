import type { ImportInfo } from "../types";

const consoleOutput = async (imports: ImportInfo[]) => {
    imports.forEach((i) => {
        const indentation = i.level! > 1 ? "│ ".repeat(i.level! - 1) : "";
        const branch = i.level! > 0 ? "├ " : "";
        const name = i.absolutePath || i.import;
        console.log(`${indentation}${branch}${name}`);
        consoleOutput(i.imports);
    });
};

export const consoleOutputPlugin = () => consoleOutput;
