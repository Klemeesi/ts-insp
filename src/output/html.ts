import * as fs from "fs";
import { ImportInfo } from "../types";

export const generateHtmlPage = (
  importInfos: ImportInfo[],
  template: string,
  save?: boolean
): string => {
  const templateContent = fs.readFileSync(template, "utf-8");
  const htmlContent = templateContent.replace(
    "<!-- IMPORTS -->",
    JSON.stringify(importInfos)
  );
  save && fs.writeFileSync("exports/foo.html", htmlContent);
  return htmlContent;
};
