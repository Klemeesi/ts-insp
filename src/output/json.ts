import { ImportInfo } from "../types";
import * as fs from "fs";

export const generateJson = (imports: ImportInfo[]) => {
    fs.writeFileSync("exports/output.json", JSON.stringify(imports, undefined, 2));
};
