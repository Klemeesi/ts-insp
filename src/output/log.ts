export const log = (...data: any[]) => console.debug("ts-insp >", ...data);
