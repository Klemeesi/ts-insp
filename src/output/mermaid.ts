import type { ImportInfo, MermaidFormatOptions } from "../types";
import * as fs from "fs";
import path from "path";
import { mermaidRenderer } from "../treeToMermaid/renderer";
import { getDefaultProcessors } from "../treeToMermaid/defaultProcessors";
import { MermaidToken } from "../treeToMermaid/types";
import { runMermaidCli } from "../treeToMermaid/cliWrapper";

const defaultOptions: MermaidFormatOptions = {
    dir: "LR",
    outputPath: "exports",
    outputName: "export",
    chartType: "graph",
    extractGroupName: (node: ImportInfo) => {
        const ap = node.absolutePath || "";
        if (ap.startsWith("node_modules") || node.type === "Unknown") {
            return "node_modules";
        } else {
            const splitted = ap.split("/");
            return splitted[splitted.length - 2] || undefined;
        }
    },
};

export const mermaidOutputPlugin = (options: MermaidFormatOptions) => async (imports: ImportInfo[]) => {
    const opt = { ...defaultOptions, ...(options || {}) };
    const output = path.resolve(opt.outputPath!, `${opt.outputName}.md`);

    const processor = mermaidRenderer.create<ImportInfo>();
    const defaultProcessors = getDefaultProcessors(opt);
    Object.keys(defaultProcessors).forEach((key) => {
        const name = key as keyof MermaidToken["type"];
        processor.register(name, defaultProcessors[key]);
    });

    const result = processor.render(imports[0], {
        extractChildren: (i) => i.imports,
    });
    fs.writeFileSync(output, result, "utf8");

    if (opt.cliOptions) {
        await runMermaidCli(opt.cliOptions, output);
    }
};
