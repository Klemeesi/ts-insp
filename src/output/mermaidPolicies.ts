import type { MermaidNodeNamePolicy, MermaidSubgraphPolicy, ResultTreeNode } from "../types";

const mermaidSubgraphPolicies = {
    none: () => undefined,
    default: (node: ResultTreeNode) => {
        const ap = node.absolutePath || "";
        if (ap.startsWith("node_modules") || node.type === "Unknown") {
            return "node_modules";
        } else {
            const splitted = ap.split("/");
            return splitted[splitted.length - 2] || undefined;
        }
    },
};

const mermaidNodeNamePolicy = {
    default: (node: ResultTreeNode) => node.moduleName,
};

export const getMermaidGroupNamePolicy = (policy?: MermaidSubgraphPolicy) => {
    if (typeof policy === "function") {
        return policy;
    }
    return (policy && mermaidSubgraphPolicies[policy]) ?? mermaidSubgraphPolicies.default;
};

export const getMermaidNodeNamePolicy = (policy?: MermaidNodeNamePolicy) => {
    if (typeof policy === "function") {
        return policy;
    }
    return (policy && mermaidNodeNamePolicy[policy]) ?? mermaidNodeNamePolicy.default;
};
