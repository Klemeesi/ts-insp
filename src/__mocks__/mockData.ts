import { ResultTreeNode } from "../types";

export const testDataSet1: ResultTreeNode[] = [
    {
        id: "root",
        uniqueId: "root",
        import: "test-data/test1.ts",
        type: "Source file",
        extension: "",
        alreadyTraversed: false,
        moduleName: "test-data/test1.ts",
        resolved: true,
        level: 0,
        properties: {},
        imports: [
            {
                id: "fs",
                uniqueId: "2aa9379e-05e2-4f51-9e2d-a0e73483f349",
                import: "fs",
                extension: "",
                alreadyTraversed: false,
                type: "Unknown",
                moduleName: "fs",
                resolved: false,
                imports: [],
                level: 1,
                properties: {},
            },
            {
                id: "src/index.ts",
                uniqueId: "882d8333-b40a-43a5-a291-2f3e2cca2dc5",
                import: "../src/index",
                extension: ".ts",
                alreadyTraversed: false,
                fullPath: "C:\\Source\\ts-insp\\src\\index.ts",
                type: "Source file",
                moduleName: "index",
                resolved: true,
                absolutePath: "src/index.ts",
                imports: [
                    {
                        id: "src/traversal.ts",
                        uniqueId: "38e2dc19-200d-4735-b2b1-e3faae42b316",
                        import: "./traversal",
                        extension: ".ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\traversal.ts",
                        type: "Source file",
                        moduleName: "traversal",
                        resolved: true,
                        absolutePath: "src/traversal.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/output/console.ts",
                        uniqueId: "2e4c29b0-4f37-4e6f-939c-afff08ba0fc0",
                        import: "./output/console",
                        extension: ".ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\output\\console.ts",
                        type: "Source file",
                        moduleName: "console",
                        resolved: true,
                        absolutePath: "src/output/console.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/commandLine.ts",
                        uniqueId: "73c2f616-e4b1-49e5-83a7-915fe3c510a0",
                        import: "./commandLine",
                        extension: ".ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\commandLine.ts",
                        type: "Source file",
                        moduleName: "commandLine",
                        resolved: true,
                        absolutePath: "src/commandLine.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/output/html.ts",
                        uniqueId: "61d1cd4b-c285-4a66-a142-d5f346617ddc",
                        import: "./output/html",
                        extension: ".ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\output\\html.ts",
                        type: "Source file",
                        moduleName: "html",
                        resolved: true,
                        absolutePath: "src/output/html.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/output/json.ts",
                        uniqueId: "c64ff942-7304-4cb1-98b9-6522d0613529",
                        import: "./output/json",
                        extension: ".ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\output\\json.ts",
                        type: "Source file",
                        moduleName: "json",
                        resolved: true,
                        absolutePath: "src/output/json.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/output/png.ts",
                        uniqueId: "1b1d7377-a84f-4ba5-9041-7900577d9c7b",
                        import: "./output/png",
                        extension: ".ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\output\\png.ts",
                        type: "Source file",
                        moduleName: "png",
                        resolved: true,
                        absolutePath: "src/output/png.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/types.d.ts",
                        uniqueId: "816f768f-efbc-4f09-a797-bae9ff22f816",
                        import: "./types",
                        extension: ".d.ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\types.d.ts",
                        type: "Source file",
                        moduleName: "types",
                        resolved: true,
                        absolutePath: "src/types.d.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                ],
                level: 1,
                properties: {},
            },
            {
                id: "src/commandLine.ts",
                uniqueId: "60917fe8-e130-4211-a3d5-530be6dcc8fd",
                import: "../src/commandLine",
                extension: ".ts",
                alreadyTraversed: false,
                fullPath: "C:\\Source\\ts-insp\\src\\commandLine.ts",
                type: "Source file",
                moduleName: "commandLine",
                resolved: true,
                absolutePath: "src/commandLine.ts",
                imports: [
                    {
                        id: "fs",
                        uniqueId: "9fe76181-9167-476e-82b1-825b3c4d6e33",
                        import: "fs",
                        extension: "",
                        alreadyTraversed: false,
                        type: "Unknown",
                        moduleName: "fs",
                        resolved: false,
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "node_modules/commander/typings/index.d.ts",
                        uniqueId: "cb973db2-f63c-497a-8169-97de3fb297ff",
                        import: "commander",
                        extension: ".d.ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\node_modules\\commander\\typings\\index.d.ts",
                        type: "Node module",
                        moduleName: "commander",
                        resolved: true,
                        absolutePath: "node_modules/commander/typings/index.d.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/types.d.ts",
                        uniqueId: "c8dfb5b0-29f7-4f3f-b1e6-891766356f68",
                        import: "./types",
                        extension: ".d.ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\types.d.ts",
                        type: "Source file",
                        moduleName: "types",
                        resolved: true,
                        absolutePath: "src/types.d.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/tsConfig.ts",
                        uniqueId: "0ee291f3-710b-42b3-9dcd-4880b7323534",
                        import: "./tsConfig",
                        extension: ".ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\tsConfig.ts",
                        type: "Source file",
                        moduleName: "tsConfig",
                        resolved: true,
                        absolutePath: "src/tsConfig.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/output/log.ts",
                        uniqueId: "02e1e976-9a0f-4ddc-9638-8eea2fc88ae5",
                        import: "./output/log",
                        extension: ".ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\output\\log.ts",
                        type: "Source file",
                        moduleName: "log",
                        resolved: true,
                        absolutePath: "src/output/log.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                ],
                level: 1,
                properties: {},
            },
            {
                id: "src/types.d.ts",
                uniqueId: "a3baf661-463f-4fd3-bb2f-97788599a802",
                import: "../src/types",
                extension: ".d.ts",
                alreadyTraversed: false,
                fullPath: "C:\\Source\\ts-insp\\src\\types.d.ts",
                type: "Source file",
                moduleName: "types",
                resolved: true,
                absolutePath: "src/types.d.ts",
                imports: [
                    {
                        id: "node_modules/typescript/lib/typescript.d.ts",
                        uniqueId: "79f85626-b66d-4648-9294-5bd13badc773",
                        import: "typescript",
                        extension: ".d.ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\node_modules\\typescript\\lib\\typescript.d.ts",
                        type: "Node module",
                        moduleName: "typescript",
                        resolved: true,
                        absolutePath: "node_modules/typescript/lib/typescript.d.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                    {
                        id: "src/output/log.ts",
                        uniqueId: "bec6f59a-040d-406e-977f-3ce77e0e8ef9",
                        import: "./output/log",
                        extension: ".ts",
                        alreadyTraversed: false,
                        fullPath: "C:\\Source\\ts-insp\\src\\output\\log.ts",
                        type: "Source file",
                        moduleName: "log",
                        resolved: true,
                        absolutePath: "src/output/log.ts",
                        imports: [],
                        level: 2,
                        properties: {},
                    },
                ],
                level: 1,
                properties: {},
            },
        ],
    },
];
