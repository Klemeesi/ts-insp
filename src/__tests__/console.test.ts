import { consoleOutput } from "../output/console";
import { testDataSet1 } from "../__mocks__/mockData";

describe("console", () => {
    it("Basic console output test", () => {
        const result = consoleOutput(testDataSet1);
        expect(result).toMatchSnapshot();
    });
});
