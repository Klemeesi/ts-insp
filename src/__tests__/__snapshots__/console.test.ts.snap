// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`console Basic console output test 1`] = `
[
  "test-data/test1.ts",
  "├ fs",
  "├ src/index.ts",
  "│ ├ src/traversal.ts",
  "│ ├ src/output/console.ts",
  "│ ├ src/commandLine.ts",
  "│ ├ src/output/html.ts",
  "│ ├ src/output/json.ts",
  "│ ├ src/output/png.ts",
  "│ ├ src/types.d.ts",
  "├ src/commandLine.ts",
  "│ ├ fs",
  "│ ├ node_modules/commander/typings/index.d.ts",
  "│ ├ src/types.d.ts",
  "│ ├ src/tsConfig.ts",
  "│ ├ src/output/log.ts",
  "│ ├ src/plugins.ts",
  "├ src/types.d.ts",
  "│ ├ node_modules/typescript/lib/typescript.d.ts",
  "│ ├ src/plugins.ts",
  "│ ├ src/output/log.ts",
  "├ src/plugins.ts",
  "│ ├ src/plugins/debugPlugin.ts",
]
`;

exports[`traversal Basic console output test 1`] = `
[
  "test-data/test1.ts",
  "├ fs",
  "├ src/index.ts",
  "│ ├ src/traversal.ts",
  "│ ├ src/output/console.ts",
  "│ ├ src/commandLine.ts",
  "│ ├ src/output/html.ts",
  "│ ├ src/output/json.ts",
  "│ ├ src/output/png.ts",
  "│ ├ src/types.d.ts",
  "├ src/commandLine.ts",
  "│ ├ fs",
  "│ ├ node_modules/commander/typings/index.d.ts",
  "│ ├ src/types.d.ts",
  "│ ├ src/tsConfig.ts",
  "│ ├ src/output/log.ts",
  "│ ├ src/plugins.ts",
  "├ src/types.d.ts",
  "│ ├ node_modules/typescript/lib/typescript.d.ts",
  "│ ├ src/plugins.ts",
  "│ ├ src/output/log.ts",
  "├ src/plugins.ts",
  "│ ├ src/plugins/debugPlugin.ts",
]
`;
